### PR TITLE
Transport instance removal

### PR DESCRIPTION
**Pull request by pepone**
This PR removes TransportInstance and related types, it also refactors the endpoint classes to not use delegation

### CODE REVIEW

File: csharp/src/Ice/TcpEndpoint.cs [hunk 1]
```diff
@@ -184,33 +167,38 @@ internal TcpEndpoint GetEndpoint(TcpAcceptor acceptor)
...
+            int port = acceptor.EffectivePort();
+            if (port == Port)
+            {
+                return this;
+            }
+            else
+            {
+                return (TcpEndpoint)CreateEndpoint(Host, port, ConnectionId, HasCompressionFlag, Timeout);
+            }
+        }
```
Comment (bentoi) [inline csharp/src/Ice/TcpEndpoint.cs:181]: Should this be a `NewPort` method instead that just take the port number?

**Comment by pepone** [comment]
> Can't we also just get rid of the WSAcceptor/SslAcceptor and WSConnector/SslConnector classes?

Don't we need those to create the WSTransceiver/SslTransceiver?

**Comment by bentoi** [comment]
We do but I suspect the TcpAcceptor/TcpConnector could create the transceiver using a new CreateTransceiver endpoint method instead?

File: csharp/src/Ice/Communicator.cs [hunk 2]
```diff
@@ -572,14 +572,9 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
...
-                    new TransportInstance(this, EndpointType.TCP, "tcp", false)));
-                AddEndpointFactory(new UdpEndpointFactory(
-                    new TransportInstance(this, EndpointType.UDP, "udp", false)));
-                AddEndpointFactory(new WSEndpointFactory(
-                    new TransportInstance(this, EndpointType.WS, "ws", false), EndpointType.TCP));
-                AddEndpointFactory(new WSEndpointFactory(
-                    new TransportInstance(this, EndpointType.WSS, "wss", true), EndpointType.SSL));
+                IceAddEndpointFactory(new TcpEndpointFactory(this));
+                IceAddEndpointFactory(new UdpEndpointFactory(this));
+                IceAddEndpointFactory(new WSEndpointFactory(this));
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator.cs:577]: And presumably the ssl and wss endpoint factories are added when the IceSSL plugin is loaded?

A follow-up PR should further integrate IceSSL in the core:
 - load ssl/wss immediately
 - eliminate the IceSSL prefix in configuration

File: csharp/src/Ice/Endpoint.cs [hunk 3]
```diff
@@ -59,11 +59,10 @@ public abstract class Endpoint : IEquatable<Endpoint>
...
         public abstract int Timeout { get; }
 
-        /// <summary>The EndpointType of this endpoint.</summary>
+        /// <summary>The <see cref="EndpointType">type</see> of this endpoint.</summary>
         public abstract EndpointType Type { get; }
-
-        /// <summary>The underlying endpoint.</summary>
-        public virtual Endpoint? Underlying => null;
+        /// <summary>The endpoint's transport unique name.</summary>
+        public abstract string Transport { get; }
```
Comment (bernardnormier) [inline csharp/src/Ice/Endpoint.cs:65]: Should we provide a default implementation for this property?
`=> Type.ToString().ToLowerInvariant()`

File: csharp/src/Ice/IPEndpoint.cs [hunk 4]
```diff
@@ -156,27 +153,23 @@ public override void IceWritePayload(OutputStream ostr)
...
-                connectors.Add(CreateConnector(p, proxy));
-            }
-            return connectors;
-        }
+        public override Endpoint NewConnectionId(string connectionId) =>
+            connectionId == ConnectionId ? this : CreateEndpoint(Host, Port, connectionId, HasCompressionFlag,
+                Timeout);
 
-        public override async ValueTask<IEnumerable<IConnector>> ConnectorsAsync(EndpointSelectionType endptSelection)
+        public override async ValueTask<IEnumerable<IConnector>> ConnectorsAsync(
```
Comment (bernardnormier) [inline csharp/src/Ice/IPEndpoint.cs:164]: A number of these methods are Ice-internal but publicly visible and should get an Ice prefix (for another PR).
Or another option could be to convert IEndpoint into an Endpoint abstract base class and use protected internal methods.

File: csharp/src/Ice/OpaqueEndpoint.cs [hunk 5]
```diff
@@ -24,6 +25,7 @@ public sealed class OpaqueEndpoint : Endpoint
 
         public override int Timeout => -1;
         public override EndpointType Type { get; }
+        public override string Transport => "";
```
Comment (bernardnormier) [inline csharp/src/Ice/OpaqueEndpoint.cs:28]: Should this be "opaque" like in stringified endpoints?
Or maybe the actual transport when Type is a known type?

File: csharp/src/Ice/TcpAcceptor.cs [hunk 6]
```diff
@@ -135,8 +144,9 @@ internal TcpAcceptor(TcpEndpoint endpoint, TransportInstance instance, string ho
             }
         }
 
+        private readonly string _adapterName;
```
Comment (bernardnormier) [inline csharp/src/Ice/TcpAcceptor.cs:147]: I assume you plan to move these fields in a future PR?

Comment (pepone) [reply csharp/src/Ice/TcpAcceptor.cs:147]: Yes we can cleanup this in a future PR, easier to review that way

File: csharp/src/Ice/Communicator.cs (hunk 2 above)
Comment (pepone) [reply csharp/src/Ice/Communicator.cs:577]: Yes currently ssl/wss are loaded when IceSSL plug-in is initialized, I assume we will also get rid of IceSSL plug-in then

File: csharp/src/Ice/Endpoint.cs (hunk 3 above)
Comment (pepone) [reply csharp/src/Ice/Endpoint.cs:65]: Actually we have Name that does the same, maybe we don't need `Transport`

File: csharp/src/Ice/IPEndpoint.cs (hunk 4 above)
Comment (pepone) [reply csharp/src/Ice/IPEndpoint.cs:164]: We need to cleanup accessibility in this API, and ensure that creating a new Transport outside is doable

File: csharp/src/Ice/OpaqueEndpoint.cs (hunk 5 above)
Comment (pepone) [reply csharp/src/Ice/OpaqueEndpoint.cs]: I removed Transport, we can just use `Name` that already return "opaque"

File: csharp/src/Ice/TcpEndpoint.cs (hunk 1 above)
Comment (pepone) [reply csharp/src/Ice/TcpEndpoint.cs]: Fixed

File: csharp/src/Ice/WSEndpoint.cs [hunk 7]
```diff
@@ -75,89 +50,111 @@ public override string OptionsToString()
...
+            new WSEndpoint(Communicator, host, port, SourceAddress, timeout, connectionId, compressionFlag, Resource);
+    }
 
-        public override Endpoint NewCompressionFlag(bool compressionFlag) =>
-            compressionFlag == _delegate.HasCompressionFlag ? this :
-                new WSEndpoint(_instance, _delegate.NewCompressionFlag(compressionFlag), Resource);
+    public sealed class WSSEndpoint : SslEndpoint
+    {
+        /// <summary> The resource of the WebSocket endpoint.</summary>
+        // TODO: better description
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/WSEndpoint.cs:106]: Fixed

**Pull request merged**